### DISCUSSION TITLE
BL-Touch and 3D Touch

### DISCUSSION THREAD

**Discussion by sidecircle-dev** [fra589/grbl-Mega-5X#206] [Q&A] [answered]
Bonjour Gauthier,

je souhaiterai utiliser une sonde 3D-Touch ou BL-Touch pour le palpage, ou puis-je trouver des instructions ?
Pour le branchement j'ai trouvé un schéma ici https://reprap.org/forum/read.php?110,779943 mais je n'arrive pas à aller plus loin, je souhaitais utiliser la fonction heightmap dans grblCtrl par exemple.

I'd like to use a 3D-Touch or BL-Touch to level, where could I find some documentation ?
I found how to connect the probe, (https://reprap.org/forum/read.php?110,779943) but then I'm stuck: how could I use the heightmap tool in grblCtrl ?

Merci beaucoup ar avance pour l'aide! Thanks a lot in advance for your valuable help
Pascal

**Comment by fra589** [comment]
Bonjour Pascal,  
  
Les sondes BL-Touch sont généralement utilisées par les imprimantes 3D et nécessitent une action supplémentaires pour activer le palpage : Activation d'une sortie PWM similaire au pilotage d'un servo destinée à sortir le palpeur pour qu'il soit plus bas que la tête d'impression.  
Cette fonctionnalité n'est pas supportée par Grbl car normalement inutile sur une fraiseuse ou un routeur. Il n'y a pas de gestion d'un port pour piloter un servo dans Grbl.  
En général, on réalise le palpage directement avec la fraise, a condition que la matière à palper soit conductrice on connecte la fraise au port probe (cf. https://github.com/fra589/grbl-Mega-5X/wiki/grbl-Mega-5X-pinout) et la pièce à palper à la masse. Ou alors on utilise une sonde zprobe (https://www.youtube.com/watch?v=3xgIS_IEX-Y) ou bien une sonde probe multidirectionnelle (https://www.instructables.com/CNC-Manual-Touch-Probe/).  
  
---
BL-Touch probes are generally used by 3D printers and require an additional action to activate probing: Activation of a PWM output similar to driving a servo intended to output the probe so that it is lower than the head printing.  
This feature is not supported by Grbl as it is normally unnecessary on a milling machine or router. There is no management of a port to drive a servo in Grbl.  
In general, the probing is carried out directly with the cutter, provided that the material to be probed is conductive, the cutter is connected to the probe port (cf. https://github.com/fra589/grbl-Mega-5X/wiki/grbl-Mega-5X-pinout) and the workpiece to the ground. Or we use a zprobe probe (https://www.youtube.com/watch?v=3xgIS_IEX-Y) or a multidirectional probe probe (https://www.instructables.com/CNC-Manual-Touch-Probe /).   
  
@++;  
Gauthier.

**Reply by sidecircle-dev** [reply]
Bonjour Gauthier,

Merci pour votre réponse rapide et détaillée. C'est bien là un peu tout mon désespoir: ma matière à palper n'est pas conductrice - je souhaite faire du dessin sur un support en toile de grande taille et forcément, la toile a des différences d'altitudes conséquentes.
J'ai trouvé grbl5x idéal car j'ai réussi à gérer l'élévation et inclinaison du pinceau. Il n'y a pas d'espoir de convertir une fonctionnalité PWM inutilisée dans grbl5x pour arriver à mes fins? 

Thanks for your detailed and fast answer. That's exactly my point: the surface I want to map isn't conductor, therefore I was looking for some BL/3D Touch solution. I'm going to apply paint on a canvas and it is all but even. I found grbl5x to be a perfect solution because it gave me a way to control the height and orientation of the pen with the painting or ink. 
Do you believe there's a way to tweek something in the software to make use of a BLTouch probe anyhow?

@#;
Pascal

**Comment by fra589** [comment, marked as answer]
Bonjour Pascal,  
  
A mon avis, la meilleur solution consisterait à utiliser une sonde de palpage "touch probe" dans la langue de Shakespeare. C'est la méthode "Pro" de répondre à ton besoin. On en trouve sur le net, c'est assez cher mais avec un peu d'habileté, on peux la fabriquer soi même cf. [google_search://"diy cnc probe"](https://www.google.com/search?client=firefox-b-d&q=diy+cnc+probe).  
  
Pour du palpage uniquement en Z, on doit pouvoir fabriquer quelque chode de simple avec simplement un switch en Z, par example un capteur de fin de course monté à la place de l'outil...  
On doit pouvoir également modifier un BLTouch pour qu'il soit sorti en permanance... Ajouter un petit ressort dedans ? ou appliquer une tension permanente au signal servo ?... A étudier et vérifier.  

Autrement, j'ai dans mes TODO l'ajout de fonctions auxiliaires tel que la commande de sorties supplémentaires [M62-M65](http://linuxcnc.org/docs/2.8/html/gcode/m-code_fr.html#sec:M62-a-M65-Ctrl-Sortie-Numerique), mais ça ne seras pas pour tout de suite... Peut être d'ici la fin de l'année... J'ai beaucoup d'autre choses dans la liste a faire :-)

@++;
Gauthier

**Comment by MaWa53** [comment]
Hello I just found this discussion and woukld like to report what I did with Bl Touch on my laser cutter.

 I am currently upgrading from grbl 1.1h running on a 328p MKS clone board delivered with my AtomStack A5.
 
I added a z Axis and wanted to drive a Bl Touch (the original one from Korea). Due to the fact that no 16 bit PWM is free on the 328p i added an Arduino leonardo which is connected to two free Ax pins. 
A small .ino code reads these two pins and "translates" them into the four servo pulse widths (deploy = 647, Alarm release=2194, Stowup=1473 and Test=17829) 
The Bl Touch is mounted on the left side of the laser unit and triggers at a z-level 2 mm below the bottom of the laser (you usually set the laser 2mm offset using a 2mm plastic card  a screw button.  
You connect the Bl Touch output to your regular z probe input. The Z axis has a Z-Home limit switch and is homed first. Just like a spindle. 

To use it you insert the material you want to cut or engrave. 
Using lightBurn you have designed you object and get X&Y into position. 
Then you issue a probe G code and start the object stream. 
To get it to work I added a BlTouch.c with the necessary code to set the two Ax output pins.
I inserted some calls into the existing probe.c. and it works very good. 

I finally ran into the code size limits of the 328p with some further additions so I am currently upgrading to an MKS Gen_L board with TMC2209 stepper drivers and Marlin( _sorry Gauthier :-_)  because I want to turn on the air_assist, evacuation fan and some leds lights in the g code and these functions including Bl Touch and UART TMC settings are readily available in Marlin as well as the mesh bed leveling I need to be able to compensate slightly bent material. 

But I have a CNC 3018 ProVer from SainSmart. Unfortunately they do not want to disclose their modifications to their 1.1f firmware and that firmware / hardware has some bugs that annoy me. And the milling software I intend to use expects a grbl controller and not a Marlin controller.
Therefore I am setting up a second MKS Gen_L **WITH Gauthier's 1.2d** but only with 3 axis and an electric Z probe (via Opto Coupler) to be able to mill prototype PCB boards. 

I intend to transfer some of the modifications (but not BLTouch) from my 1.1h upgrade to 1.2d in very near future and see how the result works.

For example an extended homing which evaluates the  max_travel settings using the - and + limit switches. It first does a normal homing as set by the settings. Then it does a fake homing in the opposite directions and using the step counts used for this move plus homing_pulloffs the max_travel values ar computed and set.

Also using separate homing_pulloff values for X,Y and Z to get the machine coordinates into a save region without have to set that region using G Code from every PC Program I use (LightBurn vs laserGRBL for example).  

And some ideas more....
The mega 2560 has plenty of space and I am a retired computer scientist and confessing addicted programmer have plenty of time. 
When I have something ready to share I will certainly issue a pull request so Gautier can decide if  "Les changements et les extensions sont à son goût".

Manfred

**Comment by fra589** [comment]
Hi Manfred,  
  
I think Marlin is not optimal for CNC milling but I'm not offended :-).  
  
The air_assist, evacuation can be driver by the new digital output (M62 to M65 GCode, grbl-Mega-5X >= v1.2a).  
grbl-Mega-5X v1.2a and upper have the ability to drive up to 4 digital outputs!  
So, today, it's possible to use a digital output to drive another microcontroller which will drive the BL-Touch or a servo.
Currently, I'm writing a new Grbl's functionality for [M67/M68 GCode function for driving analog output](http://linuxcnc.org/docs/2.8/html/gcode/m-code.html#mcode:m67). Those functions will drive a supplemental PWM output which can be used for BL-Touch or servo.  
  
I think it's not normal that SainSmart does not want to disclose its modifications to its 1.1f firmware, it seems to me a violation of the GNU GPLv3 license!  
  
Unlike your extended homing feature, which seems useless to me and at the opposite of Grbl's optimization policy (it takes up space and computation time for something that is done once and for all when configuring the machine , and therefore useless after this configuration), the possibility of separate homing_pulloff seems to me quite interesting and useful without increasing processing times.  
  
@++;  
Gauthier.

**Reply by HansdeKoster1961** [reply]
Hi , is this topic still active ? used your mega 5 fork for a laser setup as a test in combination with Lightburn , made my own Z axis mount.
Works like charm , added also a roller for the A axis , but trying to implement a 3d printer bl / 3d / cr touch .

Any progress on the M67/68 ?

grtz from Holland.

**Reply by fra589** [reply]
Hello @HansdeKoster1961,

The G codes M67 and M68 (analog outputs) have been added in grbl-Mega-5X version 1.2g (June 2024).
I have not yet taken the time to add these 2 codes in the wiki (https://github.com/fra589/grbl-Mega-5X/wiki/G-code-Quick-Reference).
The activation and configuration of this feature can be found in config.h, lines 155 to 173.

The latest version of grbl-Mega-5X is 1.2h.

@++;
Gauthier.
